src/lib/updating-element.ts

```diff
@@ -0,0 +1,501 @@
+    }
+    this._validationState = ValidationState.Invalid;
+    // Make a new validate promise and put the resolver on the stack.
+    this._validatePromise = new Promise((resolve) => this._validateResolvers.push(resolve));
+    // Wait a tick to actually process changes (allows batching).
+    await 0;
+    // Mixin instance properties once, if they exist.
+    if (this._instanceProperties) {
+      Object.assign(this, this._instanceProperties);
+      this._instanceProperties = null;
+    }
+    // Rip off changedProps.
+    const changedProps = this._changedProps || {};
+    this._changedProps = null;
+    if (this.shouldUpdate(changedProps)) {
+      // During update, setting properties does not trigger invalidation.
+      this._validationState = ValidationState.Updating;
+      this.update(changedProps);
+      this._validationState = ValidationState.Valid;
+      // During finishUpdate, setting properties does trigger invalidation,
+      // and users may choose to await other state, like children being updated.
+      await this.finishUpdate(changedProps);
+    } else {
+      this._validationState = ValidationState.Valid;
+    }
+    // Only resolve the promise if we finish in a valid state (finishUpdate
+    // did not trigger more work).
+    if (this._validationState === ValidationState.Valid) {
+      while (this._validateResolvers.length) {
+        const resolver = this._validateResolvers.pop();
+        resolver!();
+      }
+    }
+    return this._validatePromise;
+  }
+
+  private _isPendingUpdate() {
+    return this._validationState === ValidationState.Invalid ||
+      this._validationState === ValidationState.Updating;
+  }
+
+  /**
+   * Returns a Promise that resolves when the element has finished updating.
+   */
+  get updateComplete() {
+    return this._isPendingUpdate() ? this._validatePromise : Promise.resolve();
+  }
+
+  /**
+   * Controls whether or not `update` should be called when the element invalidates.
+   * By default, this method always returns true, but this can be customized to
+   * control when to update.
+   * * @param _changedProperties changed properties with old values
+   */
+  protected shouldUpdate(_changedProperties: AnyObject): boolean {
+    return true;
+  }
+
+  /**
+   * Updates the element. This method does nothing by default and should be
+   * implemented to render and keep updated DOM in the element's root.
+   * * @param _changedProperties changed properties with old values
+   */
+  protected update(_changedProperties: AnyObject) {}
+
+  /**
+   * Finishes updating the element. This method does nothing by default and
+   * should be implemented to perform post update tasks on element DOM. This
+   * async function can await other Promises to defer the resolution of the
+   * `invalidate` and `updateComplete` Proimses.
+   * * @param _changedProperties changed properties with old values
+   */
+  protected async finishUpdate(_changedProperties: AnyObject) {}
```

Comment:
Make this abstract

Reply:
I did, but TypeScript doesn't seem to support an optional abstract method. Left it in so we can verify that.

Reply:
Yeah, looks like it'd just be optional then, with no implementation. ie, optional => abstract.